### PR TITLE
ci: use `sccache` for Rust

### PR DESCRIPTION
**Pull request by thomaseizinger**
Our caches in GitHub actions are hopelessly overflowing, plus for the Kotlin and Swift jobs, we don't seem to be doing a particularly good job at caching the build outputs because those jobs take forever.

Instead of using GitHub actions, this PR configures `sccache` for all Rust compilation commands and uses a GCP bucket to store the artifacts. This speeds up some of the builds a fair bit. Android now finishes in ~6minutes.

Apart from the self-hosted MacOS 14 runner, the Swift jobs are slow but still a lot faster than what we currently have.

Windows seems to be quite slow at compiling / fetching artefacts which is negatively impacted by this change because they now have to be fetched from the bucket.

Overall, I think this is a net-positive though and should be much easier to maintain going forward.

### CODE REVIEW

**Comment by conectado** [comment]
I saw hapsoc/fluke#118 today and I wonder if it will help the same way

**Comment by thomaseizinger** [comment]
> I saw [hapsoc/fluke#118](https://github.com/hapsoc/fluke/issues/118) today and I wonder if it will help the same way

That was my inspiration too :)

**Comment by thomaseizinger** [comment]
Let's see if we are hitting the rate limits of the GitHub actions cache. Otherwise it might be worth setting up a GCP bucket. We don't have many external contributors at the moment anyway so optimising our caching for the team works. (Forks don't have access to secrets so can't access the bucket unless we make it publicly readable?)

**Comment by jamilbk** [comment]
Merging the kotlin build fix in from main

File: rust/connlib/clients/apple/build-rust.sh
Comment (jamilbk) [inline rust/connlib/clients/apple/build-rust.sh:33]: Glad you found this!

**Review by jamilbk — APPROVED** [review]
LGTM. I'm sure it'll change again at some point, but let's see how this does.

File: .github/actions/setup-rust/action.yml
Comment (AndrewDryga) [inline .github/actions/setup-rust/action.yml:17]: We should create this bucket using Terraform @thomaseizinger

Comment (jamilbk) [reply .github/actions/setup-rust/action.yml:17]: 👍

Comment (thomaseizinger) [reply .github/actions/setup-rust/action.yml:17]: > We should create this bucket using Terraform @thomaseizinger

Yeah that makes sense. I didn't think of it to be honest but testing that would have been a pain because it doesn't get applied until merge right?

**Pull request merged**
